### PR TITLE
MEX: Constraint structure and dimension detection

### PR DESCRIPTION
**Pull request by FreyJo**
- [x] Detect constraint types.
- [x] Detect dimensions.
- [x] Set up `acados_ocp_nlp_json` only when you want to generate C code.
- [x] add `Jbx_e, lbx_e, ubx_e` to implement terminal state constraints.
- [x] unified namings `forw`, `hess`
- [x] common external function generation for MEX and templating
- [x] make tera use parametric CasADi functions
- [x] mex template: works with ERK & IRK
- [x] some fixes to tera templates
- [x] add `y_ref` to cost structure detection

### CODE REVIEW

**Comment by zanellia** [comment]
Hey @FreyJo, thanks for the great work, as usual ;) On a side note, do you think it would be relatively easy to implement this in Python too?

**Comment by FreyJo** [comment]
@zanellia yes, as we spoke already, it would be very easy to implement this function with CasADi Python
https://github.com/acados/acados/pull/493/files#diff-fa7da15737db0a3cd8de05e9ae786a07

**Comment by FreyJo** [comment]
I think, this is now in a good state to be merged.
Feedback is welcome, as always. @giaf @zanellia

File: ROADMAP.md [hunk 1]
```diff
@@ -4,6 +4,9 @@
 - [ ] propagate cost in integrator
 - [ ] restore download and compilation of OOQP
 - [ ] split ocp solve into prepare and feedback
+- [ ] stage transition functions for changing model dimensions
```
Comment (giaf) [inline ROADMAP.md]: That would be good to have :) Still have to decide if it is better to make the functions explicit, implicit or both

File: ROADMAP.md [hunk 2]
```diff
@@ -4,6 +4,9 @@
...
+- [ ] fix bghp constraints to have np != 1;
```
Comment (giaf) [inline ROADMAP.md]: In general, the p constraints should be also reimplemented to be independent of `h` instead of on top of them, i.e. to be able to have a `p` constraint and an independent `h` constraint on the same stage.

File: examples/acados_matlab_octave/pendulum_on_cart_model/minimal_example_ocp.m [hunk 3]
```diff
@@ -41,11 +41,11 @@
 end
 
 %% discretization
-N = 100;
+N = 20;
```
Comment (giaf) [inline examples/acados_matlab_octave/pendulum_on_cart_model/minimal_example_ocp.m]: Do you still see the swing up with 20 stages?

**Comment by giaf** [comment]
It looks good for me. Maybe @zanellia should check the part relative to the template, I can't say much there.

File: ROADMAP.md (hunk 2 above)
Comment (FreyJo) [reply ROADMAP.md]: I think @zanellia  said he wanted to implement that by using the identity as an outer function in the `p` constraints.
Anyway, this was just to not forget.

File: examples/acados_matlab_octave/pendulum_on_cart_model/minimal_example_ocp.m (hunk 3 above)
Comment (FreyJo) [reply examples/acados_matlab_octave/pendulum_on_cart_model/minimal_example_ocp.m]: yes, the plot doesnt look that smooth though.
I just used it to check the templating stuff with less output.

**Comment by FreyJo** [comment]
Thanks for the review @giaf .

I hope you are also fine with the last two commits,
i.e. also allowing to name `impl_dae_*` instead of `impl_ode_*`.

File: examples/acados_matlab_octave/pendulum_on_cart_model/minimal_example_ocp.m (hunk 3 above)
Comment (zanellia) [reply examples/acados_matlab_octave/pendulum_on_cart_model/minimal_example_ocp.m]: Indeed, this is what is implemented in the SCQP PR ;)

File: ROADMAP.md (hunk 2 above)
Comment (giaf) [reply ROADMAP.md]: Now that it is convex-over-nonlinear (instead of just quadratic-over-nonlinear) it is indeed possible, but having the two disjointed would still be the most clean way.

Comment (giaf) [reply ROADMAP.md]: And more consistent with the constraints `BGH`, otherwise `H` means two different things in the two constraints types.

Optionally, the current `BGHP` constraint type should be renamed `BGP`, to make clear that there is no stand-alone `H`.

**Comment by giaf** [comment]
> Thanks for the review @giaf .
> 
> I hope you are also fine with the last two commits,
> i.e. also allowing to name `impl_dae_*` instead of `impl_ode_*`.

Yes it is fine with me.

File: interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c [hunk 4]
```diff
@@ -563,15 +559,15 @@ int acados_create() {
         ocp_nlp_dims_set_constraints(nlp_config, nlp_dims, i, "nh", &nh[i]);
     }
 
-    {% if dims.npd > 0 %}
+    {%- if dims.npd > 0 %}
     for (int i = 0; i < N; i++) 
         ocp_nlp_dims_set_constraints(nlp_config, nlp_dims, i, "np", &npd[i]);
-    {% endif %}
-    {% if dims.npd_e > 0 %}
+    {%- endif %}
```
Comment (zanellia) [inline interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c]: why did you add all the -s? Does it improve indentation?

File: interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c [hunk 5]
```diff
@@ -641,92 +633,56 @@ int acados_create() {
         forw_vde_casadi[i].casadi_sparsity_in = &{{ model.name }}_expl_vde_forw_sparsity_in;
         forw_vde_casadi[i].casadi_sparsity_out = &{{ model.name }}_expl_vde_forw_sparsity_out;
         forw_vde_casadi[i].casadi_work = &{{ model.name }}_expl_vde_forw_work;
-        {% if ocp.dims.np < 1 %}
-        external_function_casadi_create(&forw_vde_casadi[i]);
-		{% else %}
-        external_function_param_casadi_create(&forw_vde_casadi[i], {{ocp.dims.np}});
-		{% endif %}
+        external_function_param_casadi_create(&forw_vde_casadi[i], {{dims.np}});
```
Comment (zanellia) [inline interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c]: as a general rule, we could stick to {{ <something> }}, with spaces after {{ and before  }}, for better readability.

File: interfaces/acados_matlab_octave/acados_template_mex/+acados_template_mex/generate_solver_matlab.m [hunk 6]
```diff
@@ -1,93 +1,147 @@
...
+    %% check if t_renderer is available -> download if not
+    if ispc()
+        t_renderer_location = fullfile(getenv('ACADOS_INSTALL_DIR'),'bin','t_renderer.exe');
+    else
+        t_renderer_location = fullfile(getenv('ACADOS_INSTALL_DIR'),'bin','t_renderer');
     end
+    
+    if ~isfile( t_renderer_location )
+
+        message = ['\nDear acados user, we could not find the binaries of the ',...
```
Comment (zanellia) [inline interfaces/acados_matlab_octave/acados_template_mex/+acados_template_mex/generate_solver_matlab.m]: Maybe change to "t_renderer binaries not found"? Why "the" in front of "t_renderer". We could also add the URL to the GitHub releases page in the message ;)

File: interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c (hunk 4 above)
Comment (FreyJo) [reply interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c]: It is used to get less empty lines in case that this part is not rendered.

File: interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c (hunk 5 above)
Comment (FreyJo) [reply interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c]: alright.
